sql/moz-fx-data-shared-prod/udf/bits_from_offsets/udf.sql

```diff
@@ -0,0 +1,48 @@
+CREATE OR REPLACE FUNCTION udf.bits_from_offsets(offsets ARRAY<INT64>) AS (
```

Comment:
There is a somewhat similar function here: https://github.com/mozilla/bigquery-etl/blob/master/sql/moz-fx-data-shared-prod/udf/int_to_365_bits/udf.sql

If we switched to use `int_to_hex_string`, that would let us do passes of size 64 instead of size 8, and would abstract some of the logic out of this.

Reply:
Here's the implementation using `int_to_hex_string`:

```
CREATE OR REPLACE FUNCTION udf.bits_from_offsets(offsets ARRAY<INT64>) AS (
  (
    SELECT
      LTRIM(
        STRING_AGG(
          (
            SELECT
              FROM_HEX(
                udf.int_to_hex_string(
                  BIT_OR(
                    (
                      IF(
                        DIV(n - (64 * i), 64) = 0
                        AND (n - (64 * i)) >= 0,
                        1 << MOD(n - (64 * i), 64),
                        0
                      )
                    )
                  )
                )
              )
            FROM
              UNNEST(offsets) AS n
          ),
          b''
        ),
        b'\x00'
      )
    FROM
      -- Each iteration handles 8 bits, so 256 iterations gives us 2048 bits,
      -- about 5.6 years worth. This function performs an LTRIM, so additional
      -- bits are dropped and it's safe to expand this range arbitrarily.
      UNNEST(GENERATE_ARRAY(31, 0, -1)) AS i
  )
);
```

It avoids the comment about `CODE_POINTS_TO_BYTES` but otherwise ends up having exactly the same structure. My preference would be to keep the implementation with `CODE_POINTS_TO_BYTES` as I suspect it leads to lower query complexity on the BQ side (it avoids a nested UDF call and an additional level of nested aggregation inside that UDF).

Reply:
I agree, there is an extra layer inside there. Thanks for writing this out.